crates/rpc/rpc/src/eth/sim_bundle.rs

```diff
@@ -21,20 +72,409 @@ impl<Eth> EthSimBundle<Eth> {
+                let db = CacheDB::new(StateProviderDatabase::new(state));
+
+                let initial_coinbase_balance = DatabaseRef::basic_ref(&db, coinbase)
+                    .map_err(EthApiError::from_eth_err)?
+                    .map(|acc| acc.balance)
+                    .unwrap_or_default();
+
+                let mut coinbase_balance_before_tx = initial_coinbase_balance;
+                let mut total_gas_used = 0;
+                let mut total_profit = U256::ZERO;
+                let mut refundable_value = U256::ZERO;
+                let mut body_logs: Vec<SimBundleLogs> = Vec::new();
+
+                let mut evm = RpcNodeCore::evm_config(&eth_api).evm_with_env(db, env);
+
+                for (i, item) in flattened_bundle.iter().enumerate() {
+                    // Check inclusion constraints
+                    let block_number = item.inclusion.block_number();
+                    let max_block_number =
+                        item.inclusion.max_block_number().unwrap_or(block_number);
+
+                    if current_block_number < block_number ||
+                        current_block_number > max_block_number
+                    {
+                        return Err(EthApiError::InvalidParams(
+                            EthSimBundleError::InvalidInclusion.to_string(),
+                        )
+                        .into());
+                    }
+                    RpcNodeCore::evm_config(&eth_api).fill_tx_env(
+                        evm.tx_mut(),
+                        &item.tx,
+                        item.signer,
+                    );
+
+                    let ResultAndState { result, state } =
+                        evm.transact().map_err(EthApiError::from_eth_err)?;
+
+                    if !result.is_success() && !item.can_revert {
+                        return Err(EthApiError::InvalidParams(
+                            EthSimBundleError::BundleTransactionFailed.to_string(),
+                        )
+                        .into());
+                    }
+
+                    let gas_used = result.gas_used();
+                    total_gas_used += gas_used;
+
+                    // coinbase is always present in the result state
+                    let coinbase_balance_after_tx =
+                        state.get(&coinbase).map(|acc| acc.info.balance).unwrap_or_default();
+
+                    let coinbase_diff =
+                        coinbase_balance_after_tx.saturating_sub(coinbase_balance_before_tx);
+                    total_profit += coinbase_diff;
+
+                    // Add to refundable value if this tx index is not in refund config
+                    let has_refund = item
+                        .validity
+                        .as_ref()
+                        .and_then(|v| v.refund.as_ref())
+                        .map(|refunds| refunds.iter().any(|r| r.body_idx as usize == i))
```

Comment:
this doesn't look right, `i` represents index of transaction in overall flattened bundle while `r.body_idx` is relative to the initial bundle this tx was submitted in

could we perhaps store the `refund_percent` directly on flattened items? given that when we resolve them we can find it in the validity list

Reply:
makes sense to store refund percent directly, will implement!